File: js/src/brokers/toss.ts
/**
 * 토스증권 Open API 어댑터 (REST v1.2.15).
 * ⚠️ 실전 전용 · 실측 전 — 샌드박스가 없다. 공식 OpenAPI 문서로 구현, 실계좌 소액 검증 전까지 "미검증". liveTradingEnabled 와 금액 상한 필수.
 * - /api/v1/… + Authorization: Bearer. 계좌·자산·주문 API 는 X-Tossinvest-Account: {accountSeq} 헤더 필수
 * - 성공 {"result": …}, 에러 {"error": {requestId, code, message}}. client 당 유효 토큰 1개(재발급 시 이전 토큰 무효)
 * - 한 계좌로 KRX·미국을 다룬다 → 보유·주문 심볼은 KRX:005930 / US:AAPL 로 접두를 붙여 돌려준다
 * - 시세에 등락·거래량 없음, 예수금 없음(KRW 매수가능금액 대체), 체결 엔드포인트 없음(종료 주문 execution 집계), 취소는 새 orderId 발급, 캘린더 KRX 합성
 */
import { randomUUID } from "node:crypto";
import { Decimal } from "decimal.js";
import { BrokerClient, RateLimiter, httpJson, krxCalendar, krxTickRound } from "../broker.js";
import {
  AuthError, BrokerApiError, InsufficientFundsError, InvalidOrderError, MarketClosedError, OrderNotFoundError, RateLimitError,
} from "../errors.js";
import type {
  Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
  Fill, Holding, MarketDay, Order, OrderSide, OrderStatus, OrderType, Quote, TradingEnvironment,
} from "../models.js";
import { parseSymbol, symbolCodeFor } from "../models.js";

const STATUS: Record<string, OrderStatus> = {
  PENDING: "SUBMITTED", PENDING_REPLACE: "SUBMITTED", PENDING_CANCEL: "PENDING_CANCEL", PARTIAL_FILLED: "PARTIALLY_FILLED",
  FILLED: "FILLED", CANCELED: "CANCELED", REPLACED: "CANCELED", REJECTED: "REJECTED", CANCEL_REJECTED: "REJECTED", REPLACE_REJECTED: "REJECTED",
};
const num = (v: unknown): Decimal | null => { if (v === null || v === undefined || v === "") return null; try { return new Decimal(String(v)); } catch { return null; } };
const ts = (v: unknown): Date | null => { if (!v) return null; const d = new Date(String(v)); return Number.isNaN(d.getTime()) ? null : d; };
const krw = (v: unknown): Decimal | null => (v && typeof v === "object" ? num((v as Record<string, unknown>).krw) : num(v));

export class TossClient implements BrokerClient {
  readonly capabilities: BrokerCapabilities = {
    brokerId: "toss", market: "KRX", currency: "KRW",
    candleIntervals: new Set<CandleInterval>(["1m", "1d"]),
    clientOrderId: true, nativeBracket: false, fractionalShares: false, serverOpenOrders: true,
    environments: new Set<TradingEnvironment>(["LIVE"]), markets: new Set(["KRX", "US"]),
  };

  private token: string | null = null;
  private tokenExpiresAt = 0;
  private accountSeq: string;
  private readonly limiter: RateLimiter;

  /** accountSeq 를 비우면 GET /api/v1/accounts 의 첫 BROKERAGE 계좌를 쓴다 */
  constructor(
    private readonly clientId: string,
    private readonly clientSecret: string,
    accountSeq = "",
    readonly baseUrl: string = "https://openapi.tossinvest.com",
    throttleMs = 200,
    readonly environment: TradingEnvironment = "LIVE",
  ) {
    this.accountSeq = accountSeq;
    this.limiter = new RateLimiter(throttleMs, 3, (attempt) => 1000 * 2 ** (attempt - 1));
  }

  async getQuotes(symbols: string[]): Promise<Quote[]> {
    const requested = new Map(symbols.map((s) => [symbolCodeFor(this.capabilities, s), s]));
    const result = (await this.call("GET", `/api/v1/prices?symbols=${[...requested.keys()].join(",")}`)) as Record<string, unknown>[];
    const quotes: Quote[] = [];
    for (const q of result ?? []) {
      const price = num(q.lastPrice);
      if (!price) continue;
      quotes.push({ symbol: requested.get(String(q.symbol)) ?? String(q.symbol), price, bidPrice: null, askPrice: null, volume: 0, change: null, changeRate: null, timestamp: ts(q.timestamp) ?? new Date() });
    }
    return quotes;
  }

  async getCandles(symbol: string, interval: CandleInterval, limit?: number): Promise<Candle[]> {
    if (!this.capabilities.candleIntervals.has(interval)) throw new Error(`토스 어댑터는 1m/1d 캔들만 지원합니다 (${interval})`);
    const count = Math.max(1, Math.min(limit ?? 100, 200));
    const result = (await this.call("GET", `/api/v1/candles?symbol=${symbolCodeFor(this.capabilities, symbol)}&interval=${interval}&count=${count}`)) as Record<string, unknown>;
    const candles: Candle[] = [];
    for (const c of ((result?.candles ?? []) as Record<string, unknown>[])) {
      const t = ts(c.timestamp);
      if (!t) continue;
      candles.push({ timestamp: t, open: num(c.openPrice) ?? new Decimal(0), high: num(c.highPrice) ?? new Decimal(0), low: num(c.lowPrice) ?? new Decimal(0), close: num(c.closePrice) ?? new Decimal(0), volume: Number(num(c.volume) ?? 0) });
    }
    candles.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
    return candles;
  }

  async getCalendar(): Promise<MarketDay[]> { return krxCalendar(); }

  async getAccount(): Promise<Account> {
    const cash = await this.buyingPower("KRW");
    const holdings = await this.getHoldings();
    const marketValue = holdings.reduce((acc, h) => acc.plus(h.marketValue ?? 0), new Decimal(0));
    return { accountId: await this.account(), currency: "KRW", cash, portfolioValue: cash.plus(marketValue), status: "ACTIVE", name: null };
  }

  async getHoldings(): Promise<Holding[]> {
    const result = (await this.call("GET", "/api/v1/holdings", undefined, true)) as Record<string, unknown>;
    const holdings: Holding[] = [];
    for (const h of ((result?.items ?? []) as Record<string, unknown>[])) {
      const quantity = num(h.quantity);
      if (!quantity || quantity.lte(0)) continue;
      const mv = (h.marketValue ?? {}) as Record<string, unknown>;
      const pl = (h.profitLoss ?? {}) as Record<string, unknown>;
      holdings.push({
        symbol: `${h.marketCountry === "US" ? "US" : "KRX"}:${h.symbol}`, quantity, avgEntryPrice: num(h.averagePurchasePrice) ?? new Decimal(0),
        currentPrice: num(h.lastPrice), marketValue: krw(mv.amount), unrealizedPnl: krw(pl.amount), unrealizedPnlRate: num(pl.rate),
      });
    }
    return holdings;
  }

  async getBuyingPower(): Promise<Decimal> { return this.buyingPower("KRW"); }

  async createOrder(request: CreateOrderRequest): Promise<Order> {
    const { market } = parseSymbol(request.symbol);
    const code = symbolCodeFor(this.capabilities, request.symbol);
    const isKrx = (market ?? "KRX") === "KRX";
    const isLimit = request.orderType === "LIMIT";
    const body: Record<string, unknown> = { clientOrderId: request.clientOrderId || randomUUID(), symbol: code, side: request.side, orderType: request.orderType, timeInForce: "DAY", quantity: request.quantity.toString(), confirmHighValueOrder: false };
    if (isLimit) body.price = isKrx ? krxTickRound(request.limitPrice!).toString() : request.limitPrice!.toString();
    const result = (await this.call("POST", "/api/v1/orders", body, true)) as Record<string, unknown>;
    const orderId = String(result?.orderId ?? "").trim();
    if (!orderId) throw new BrokerApiError(200, null, "토스 주문 응답에 orderId 가 없습니다");
    return { orderId, status: "SUBMITTED", symbol: `${isKrx ? "KRX" : "US"}:${code}`, side: request.side, orderType: request.orderType, quantity: request.quantity, limitPrice: request.limitPrice ?? null, filledQuantity: new Decimal(0), clientOrderId: (result.clientOrderId as string) ?? request.clientOrderId ?? null, submittedAt: new Date() };
  }

  async getOrders(): Promise<Order[]> {
    const result = (await this.call("GET", "/api/v1/orders?status=OPEN", undefined, true)) as Record<string, unknown>;
    return ((result?.orders ?? []) as Record<string, unknown>[]).map((o) => this.toOrder(o));
  }

  async getOrder(orderId: string): Promise<Order> {
    return this.toOrder((await this.call("GET", `/api/v1/orders/${orderId}`, undefined, true)) as Record<string, unknown>);
  }

  async cancelOrder(orderId: string): Promise<Order> {
    await this.call("POST", `/api/v1/orders/${orderId}/cancel`, {}, true); // 새 orderId 발급 — 원주문 ID 유지
    return { orderId, status: "PENDING_CANCEL", canceledAt: new Date() };
  }

  async getFills(): Promise<Fill[]> {
    const result = (await this.call("GET", "/api/v1/orders?status=CLOSED", undefined, true)) as Record<string, unknown>;
    const fills: Fill[] = [];
    for (const o of ((result?.orders ?? []) as Record<string, unknown>[])) {
      const ex = (o.execution ?? {}) as Record<string, unknown>;
      const qty = num(ex.filledQuantity);
      if (!qty || qty.lte(0)) continue;
      fills.push({ fillId: null, orderId: String(o.orderId), symbol: `${o.currency === "USD" ? "US" : "KRX"}:${o.symbol}`, side: (o.side as OrderSide) ?? null, quantity: qty, price: num(ex.averageFilledPrice) });
    }
    return fills;
  }

  private toOrder(o: Record<string, unknown>): Order {
    const ex = (o.execution ?? {}) as Record<string, unknown>;
    return {
      orderId: String(o.orderId ?? ""), status: STATUS[String(o.status)] ?? "UNKNOWN", symbol: `${o.currency === "USD" ? "US" : "KRX"}:${o.symbol}`,
      side: (o.side as OrderSide) ?? null, orderType: (o.orderType as OrderType) ?? null, quantity: num(o.quantity), limitPrice: num(o.price),
      filledQuantity: num(ex.filledQuantity) ?? new Decimal(0), avgFillPrice: num(ex.averageFilledPrice), clientOrderId: (o.clientOrderId as string) ?? null,
      submittedAt: ts(o.orderedAt), canceledAt: ts(o.canceledAt),
    };
  }

  private async buyingPower(currency: string): Promise<Decimal> {
    return num(((await this.call("GET", `/api/v1/buying-power?currency=${currency}`, undefined, true)) as Record<string, unknown>)?.cashBuyingPower) ?? new Decimal(0);
  }

  private async account(): Promise<string> {
    if (this.accountSeq) return this.accountSeq;
    const accounts = (await this.call("GET", "/api/v1/accounts")) as Record<string, unknown>[];
    const picked = accounts.find((a) => a.accountType === "BROKERAGE") ?? accounts[0];
    if (!picked) throw new BrokerApiError(200, null, "토스 계좌 목록이 비어 있습니다");
    this.accountSeq = String(picked.accountSeq);
    return this.accountSeq;
  }

  private call(method: string, path: string, body?: Record<string, unknown>, account = false): Promise<unknown> {
    return this.limiter.execute(() => this.callOnce(method, path, body, account), `toss ${path}`);
  }

  private async callOnce(method: string, path: string, body?: Record<string, unknown>, account = false): Promise<unknown> {
    const headers: Record<string, string> = { Authorization: `Bearer ${await this.getToken()}` };
    if (account) headers["X-Tossinvest-Account"] = await this.account();
    if (body) headers["Content-Type"] = "application/json";
    const [status, parsed, resHeaders] = await httpJson(this.baseUrl + path, { method, headers, body: body ? JSON.stringify(body) : undefined });
    if (status < 200 || status >= 300) {
      const error = (parsed.error ?? {}) as Record<string, unknown>;
      const code = String(error.code ?? "");
      const msg = `Toss(${path}) [${code}] ${error.message ?? ""} requestId=${error.requestId ?? ""}`.trim();
      const retryAfter = Number(resHeaders?.["retry-after"]);
      if (status === 429) throw new RateLimitError(status, code, msg, Number.isFinite(retryAfter) ? retryAfter : null);
      if (status === 401) throw new AuthError(status, code, msg);
      if (code === "order-not-found") throw new OrderNotFoundError(code, msg);
      if (code === "insufficient-buying-power") throw new InsufficientFundsError(status, code, msg);
      if (code === "order-hours-closed") throw new MarketClosedError(status, code, msg);
      if (status === 400 || status === 409 || status === 422) throw new InvalidOrderError(status, code, msg);
      throw new BrokerApiError(status, code, msg);
    }
    return parsed.result;
  }

  private async getToken(): Promise<string> {
    if (this.token && Date.now() < this.tokenExpiresAt - 60_000) return this.token;
    await this.limiter.throttle.wait();
    const form = new URLSearchParams({ grant_type: "client_credentials", client_id: this.clientId, client_secret: this.clientSecret });
    const [status, body] = await httpJson(`${this.baseUrl}/oauth2/token`, { method: "POST", headers: { "Content-Type": "application/x-www-form-urlencoded" }, body: form.toString() });
    if (status !== 200 || typeof body.access_token !== "string") throw new AuthError(status, (body.error as string) ?? null, `토스 토큰 발급 실패(${body.error}): ${body.error_description ?? ""}`);
    this.token = body.access_token;
    this.tokenExpiresAt = Date.now() + Number(body.expires_in ?? 86400) * 1000;
    return this.token;
  }
}

File: python/hermetix/brokers/toss.py
"""토스증권 Open API 어댑터 (REST v1.2.15).

⚠️ 실전 전용 · 실측 전 — 토스증권은 모의투자 샌드박스가 없다. 공식 OpenAPI 문서로 구현했고 실계좌 소액 검증 전까지 "미검증".
반드시 live_trading_enabled=True 와 주문 금액 상한을 설정하고 소액으로 시작하라.

규약: /api/v1/… + Authorization: Bearer. 계좌·자산·주문 API 는 X-Tossinvest-Account: {accountSeq} 헤더 필수.
성공 {"result": …}, 에러 {"error": {requestId, code, message, data}}. client 당 유효 토큰 1개(재발급 시 이전 토큰 무효).
한 계좌로 KRX·미국을 다룬다 → 보유·주문 심볼은 KRX:005930 / US:AAPL 로 접두를 붙여 돌려준다.
공통 모델과의 차이: 시세에 등락·거래량 없음, 예수금 없음(KRW 매수가능금액으로 대체), 체결 엔드포인트 없음(종료 주문 execution 집계),
취소·정정은 새 orderId 발급(원주문 ID 유지 + PENDING_CANCEL), 캘린더는 KRX 합성.
"""
from __future__ import annotations

import threading
import time
import uuid
from datetime import datetime, timezone
from decimal import Decimal

from ..broker import BrokerClient, RateLimiter, _Http, krx_calendar, krx_tick_round
from ..errors import (
    AuthError, BrokerApiError, InsufficientFundsError, InvalidOrderError, MarketClosedError, OrderNotFoundError, RateLimitError,
)
from ..models import (
    Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
    Fill, Holding, MarketDay, Order, OrderSide, OrderStatus, OrderType, Quote, TradingEnvironment, parse_symbol,
)

_STATUS = {
    "PENDING": OrderStatus.SUBMITTED, "PENDING_REPLACE": OrderStatus.SUBMITTED, "PENDING_CANCEL": OrderStatus.PENDING_CANCEL,
    "PARTIAL_FILLED": OrderStatus.PARTIALLY_FILLED, "FILLED": OrderStatus.FILLED,
    "CANCELED": OrderStatus.CANCELED, "REPLACED": OrderStatus.CANCELED,
    "REJECTED": OrderStatus.REJECTED, "CANCEL_REJECTED": OrderStatus.REJECTED, "REPLACE_REJECTED": OrderStatus.REJECTED,
}


def _d(value) -> Decimal | None:
    if value is None or value == "":
        return None
    try:
        return Decimal(str(value))
    except Exception:  # noqa: BLE001
        return None


def _ts(value) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except ValueError:
        return None


class TossClient(BrokerClient):

    capabilities = BrokerCapabilities(
        broker_id="toss", market="KRX", currency="KRW",
        candle_intervals=frozenset({CandleInterval.MIN_1, CandleInterval.DAY_1}),
        client_order_id=True, native_bracket=False, fractional_shares=False, server_open_orders=True,
        environments=frozenset({TradingEnvironment.LIVE}),  # 샌드박스 없음
        markets=frozenset({"KRX", "US"}),
    )

    def __init__(self, client_id: str, client_secret: str, account_seq: str = "", base_url: str = "https://openapi.tossinvest.com",
                 throttle_seconds: float = 0.2, environment: TradingEnvironment = TradingEnvironment.LIVE):
        """account_seq 를 비우면 GET /api/v1/accounts 의 첫 BROKERAGE 계좌를 쓴다."""
        self.environment = environment
        self._client_id = client_id
        self._client_secret = client_secret
        self._account_seq = account_seq
        self._http = _Http(base_url)
        self._limiter = RateLimiter(throttle_seconds, max_retries=3, backoff=lambda attempt: float(1 << (attempt - 1)))
        self._token: str | None = None
        self._token_expires_at = 0.0
        self._token_lock = threading.Lock()

    # ------------------------------------------------------------------ market

    def get_quotes(self, symbols: list[str]) -> list[Quote]:
        requested = {self.capabilities.symbol_code(s): s for s in symbols}
        result = self._call("GET", "/api/v1/prices", query={"symbols": ",".join(requested)})
        quotes = []
        for q in result or []:
            price = _d(q.get("lastPrice"))
            if price is None:
                continue
            quotes.append(Quote(symbol=requested.get(q.get("symbol"), q.get("symbol")), price=price, bid_price=None, ask_price=None,
                                volume=0, change=None, change_rate=None,  # /prices 에 등락·거래량 없음
                                timestamp=_ts(q.get("timestamp")) or datetime.now(timezone.utc)))
        return quotes

    def get_candles(self, symbol: str, interval: CandleInterval, limit: int | None = None) -> list[Candle]:
        if interval not in self.capabilities.candle_intervals:
            raise ValueError(f"토스 어댑터는 1m/1d 캔들만 지원합니다 ({interval.value})")
        count = max(1, min(limit or 100, 200))
        result = self._call("GET", "/api/v1/candles", query={"symbol": self.capabilities.symbol_code(symbol), "interval": interval.value, "count": str(count)})
        candles = []
        for c in (result or {}).get("candles") or []:
            ts = _ts(c.get("timestamp"))
            if ts is None:
                continue
            candles.append(Candle(timestamp=ts, open=_d(c.get("openPrice")) or Decimal(0), high=_d(c.get("highPrice")) or Decimal(0),
                                  low=_d(c.get("lowPrice")) or Decimal(0), close=_d(c.get("closePrice")) or Decimal(0),
                                  volume=int(_d(c.get("volume")) or 0)))
        candles.sort(key=lambda c: c.timestamp)  # 최신순 → 과거→최신
        return candles

    def get_calendar(self) -> list[MarketDay]:
        return krx_calendar()

    # ----------------------------------------------------------------- account

    def get_account(self) -> Account:
        cash = self._buying_power("KRW")
        market_value = sum((h.market_value or Decimal(0) for h in self.get_holdings()), Decimal(0))
        return Account(account_id=self._account(), currency="KRW", cash=cash, portfolio_value=cash + market_value)

    def get_holdings(self) -> list[Holding]:
        result = self._call("GET", "/api/v1/holdings", account=True) or {}
        holdings = []
        for h in result.get("items") or []:
            qty = _d(h.get("quantity"))
            if qty is None or qty <= 0:
                continue
            market = "US" if h.get("marketCountry") == "US" else "KRX"
            holdings.append(Holding(
                symbol=f"{market}:{h.get('symbol')}", quantity=qty,
                avg_entry_price=_d(h.get("averagePurchasePrice")) or Decimal(0), current_price=_d(h.get("lastPrice")),
                market_value=_d(((h.get("marketValue") or {}).get("amount"))) if not isinstance((h.get("marketValue") or {}).get("amount"), dict) else _d((h.get("marketValue") or {}).get("amount", {}).get("krw")),
                unrealized_pnl=_d(((h.get("profitLoss") or {}).get("amount"))) if not isinstance((h.get("profitLoss") or {}).get("amount"), dict) else _d((h.get("profitLoss") or {}).get("amount", {}).get("krw")),
                unrealized_pnl_rate=_d((h.get("profitLoss") or {}).get("rate")),  # 이미 소수 비율
            ))
        return holdings

    def get_buying_power(self) -> Decimal:
        return self._buying_power("KRW")

    # ------------------------------------------------------------------ orders

    def create_order(self, request: CreateOrderRequest) -> Order:
        market, _ = parse_symbol(request.symbol)
        code = self.capabilities.symbol_code(request.symbol)
        is_krx = (market or "KRX") == "KRX"
        is_limit = request.order_type == OrderType.LIMIT
        body = {"clientOrderId": request.client_order_id or str(uuid.uuid4()), "symbol": code, "side": request.side.value,
                "orderType": request.order_type.value, "timeInForce": "DAY", "quantity": str(request.quantity), "confirmHighValueOrder": False}
        if is_limit:
            body["price"] = str(krx_tick_round(request.limit_price)) if is_krx else str(request.limit_price)
        result = self._call("POST", "/api/v1/orders", json_body=body, account=True) or {}
        order_id = str(result.get("orderId", "")).strip()
        if not order_id:
            raise BrokerApiError(200, None, "토스 주문 응답에 orderId 가 없습니다")
        return Order(order_id=order_id, status=OrderStatus.SUBMITTED, symbol=f"{'KRX' if is_krx else 'US'}:{code}", side=request.side,
                     order_type=request.order_type, quantity=request.quantity, limit_price=request.limit_price,
                     filled_quantity=Decimal(0), client_order_id=result.get("clientOrderId") or request.client_order_id,
                     submitted_at=datetime.now(timezone.utc))

    def get_orders(self) -> list[Order]:
        result = self._call("GET", "/api/v1/orders", query={"status": "OPEN"}, account=True) or {}
        return [self._to_order(o) for o in result.get("orders") or []]

    def get_order(self, order_id: str) -> Order:
        return self._to_order(self._call("GET", f"/api/v1/orders/{order_id}", account=True) or {})

    def cancel_order(self, order_id: str) -> Order:
        self._call("POST", f"/api/v1/orders/{order_id}/cancel", json_body={}, account=True)  # 새 orderId 발급 — 원주문 ID 유지
        return Order(order_id=order_id, status=OrderStatus.PENDING_CANCEL, canceled_at=datetime.now(timezone.utc))

    def get_fills(self) -> list[Fill]:
        result = self._call("GET", "/api/v1/orders", query={"status": "CLOSED"}, account=True) or {}
        fills = []
        for o in result.get("orders") or []:
            ex = o.get("execution") or {}
            qty = _d(ex.get("filledQuantity"))
            if qty is None or qty <= 0:
                continue
            side = o.get("side")
            fills.append(Fill(fill_id=None, order_id=o.get("orderId"), symbol=f"{'US' if o.get('currency') == 'USD' else 'KRX'}:{o.get('symbol')}",
                              side=OrderSide(side) if side in ("BUY", "SELL") else None, quantity=qty, price=_d(ex.get("averageFilledPrice"))))
        return fills

    # ---------------------------------------------------------------- internal

    @staticmethod
    def _to_order(o: dict) -> Order:
        ex = o.get("execution") or {}
        side, otype = o.get("side"), o.get("orderType")
        return Order(order_id=str(o.get("orderId", "")), status=_STATUS.get(str(o.get("status")), OrderStatus.UNKNOWN),
                     symbol=f"{'US' if o.get('currency') == 'USD' else 'KRX'}:{o.get('symbol')}",
                     side=OrderSide(side) if side in ("BUY", "SELL") else None,
                     order_type=OrderType(otype) if otype in ("MARKET", "LIMIT") else None,
                     quantity=_d(o.get("quantity")), limit_price=_d(o.get("price")), filled_quantity=_d(ex.get("filledQuantity")) or Decimal(0),
                     avg_fill_price=_d(ex.get("averageFilledPrice")), client_order_id=o.get("clientOrderId"),
                     submitted_at=_ts(o.get("orderedAt")), canceled_at=_ts(o.get("canceledAt")))

    def _buying_power(self, currency: str) -> Decimal:
        return _d((self._call("GET", "/api/v1/buying-power", query={"currency": currency}, account=True) or {}).get("cashBuyingPower")) or Decimal(0)

    def _account(self) -> str:
        if self._account_seq:
            return self._account_seq
        with self._token_lock:
            if self._account_seq:
                return self._account_seq
            accounts = self._call("GET", "/api/v1/accounts") or []
            picked = next((a for a in accounts if a.get("accountType") == "BROKERAGE"), accounts[0] if accounts else None)
            if picked is None:
                raise BrokerApiError(200, None, "토스 계좌 목록이 비어 있습니다")
            self._account_seq = str(picked.get("accountSeq"))
            return self._account_seq

    def _call(self, method: str, path: str, *, query: dict | None = None, json_body: dict | None = None, account: bool = False):
        return self._limiter.execute(lambda: self._call_once(method, path, query, json_body, account), f"toss {path}")

    def _call_once(self, method: str, path: str, query, json_body, account: bool):
        headers = {"Authorization": f"Bearer {self._get_token()}"}
        if account:
            headers["X-Tossinvest-Account"] = self._account()
        status, parsed = self._http.request(method, path, headers=headers, query=query, json_body=json_body)
        if status < 200 or status >= 300:
            error = parsed.get("error") or {}
            code = str(error.get("code", ""))
            msg = f"Toss({path}) [{code}] {error.get('message', '')} requestId={error.get('requestId', '')}".strip()
            retry_after = (getattr(self._http, "last_headers", None) or {}).get("retry-after")
            try:
                retry_after = float(retry_after) if retry_after else None
            except ValueError:
                retry_after = None
            if status == 429:
                raise RateLimitError(status, code, msg, retry_after)
            if status == 401:
                raise AuthError(status, code, msg)
            if code == "order-not-found":
                raise OrderNotFoundError(code, msg)
            if code == "insufficient-buying-power":
                raise InsufficientFundsError(status, code, msg)
            if code == "order-hours-closed":
                raise MarketClosedError(status, code, msg)
            if status in (400, 409, 422):
                raise InvalidOrderError(status, code, msg)
            raise BrokerApiError(status, code, msg)
        return parsed.get("result")

    def _get_token(self) -> str:
        if self._token and time.time() < self._token_expires_at - 60:
            return self._token
        with self._token_lock:
            if self._token and time.time() < self._token_expires_at - 60:
                return self._token
            self._limiter.throttle.wait()
            status, body = self._http.request("POST", "/oauth2/token", form_body={
                "grant_type": "client_credentials", "client_id": self._client_id, "client_secret": self._client_secret})
            if status != 200 or not body.get("access_token"):
                raise AuthError(status, body.get("error"), f"토스 토큰 발급 실패({body.get('error')}): {body.get('error_description') or ''}")
            self._token = body["access_token"]
            self._token_expires_at = time.time() + float(body.get("expires_in", 86400))
            return self._token
